### ISSUE TITLE
Enum codegen - cases not public when type is internal

### ISSUE THREAD

**Issue by TIHan** [Microsoft/visualfsharp#4821]
This is from an email that @KevinRansom wrote:

"Whilst adding etw tracing to the editor Will and Brett came across an interesting wrinkle in our codegen.  The F# code generator for enums when generating code for an internal enum makes each case internal.  When C# generates code for enum for internal types it makes the case public."

C# code:
```csharp
enum MyColor { Red, Yellow, Blue };
```
C# IL Output:
```il
.class private auto ansi serializable sealed MyColor
    extends [netstandard]System.Enum
{
    .custom instance void     .field public static literal valuetype ClassLibrary10.MyColor Blue = int32(2)

    .field public static literal valuetype ClassLibrary10.MyColor Red = int32(0)

    .field public specialname rtspecialname int32 value__

    .field public static literal valuetype ClassLibrary10.MyColor Yellow = int32(1)

}
```

F# code:
```fsharp
type internal Color =
    | Red = 0
    | Yellow = 1
    | Blue = 2
```

F# IL Output:
```il
.class private auto ansi serializable sealed Color
    extends [netstandard]System.Enum
{
    .custom instance void [FSharp.Core]Microsoft.FSharp.Core.CompilationMappingAttribute::.ctor(valuetype [FSharp.Core]Microsoft.FSharp.Core.SourceConstructFlags) = { int32(3) }
    .field assembly static literal valuetype ClassLibrary1.Color Blue = int32(2)

    .field assembly static literal valuetype ClassLibrary1.Color Red = int32(0)

    .field assembly specialname rtspecialname int32 value__

    .field assembly static literal valuetype ClassLibrary1.Color Yellow = int32(1)

}
```

"In F# code this is not a problem.  However, Will discovered that when using the Trace logging facility to emit enum values to the log file, F# enums are emitted as their numeric value, whereas C# enums use the label representation.

Ie Log(Color.Blue) becomes “2” in the log.  Whereas Log(MyColor.Blue) becomes “Blue”

This occurs because the logger does not recognize cases that are not marked public."

**Comment by saul** [comment]
This is a dupe of https://github.com/Microsoft/visualfsharp/issues/2820, it's by design

**Comment by nosami** [comment]
I had a similar issue in VS for Mac. I needed to use strings for the command handler attributes [here](https://github.com/mono/monodevelop/blob/e5933809f0f3cb8917e50051bc1e4b9eacc243e9/main/external/fsharpbinding/MonoDevelop.FSharpBinding/ScriptDebugging.fs#L155) instead of strongly typed enums.

**Comment by KevinRansom** [comment]
The scenarios where this matters is reflection scenarios.  We do want to change the codegen to match C#.  However, there is a non zero chance that this will break existing reflection code.

We will be proposing a design for enabling us to make opt-in breaking changes and yet keep existing projects and solutions compiling as before.  Once we have an agreed design for that we will start being more agressive about improving codegen in ways that may break existing code.  We expect that new projects will automatically opt-in to the new behavior, and that the new behavior will be the de-facto F# codegen style.

Kevin

**Comment by dsyme** [comment]
@saul I think it's not quite the same ass #2820: it's about the codegen that gets done for Enum types by default (which I'm ok with), rather than changing the language to allow "public" members inside "internal" types (something I'm loathe to do)

**Comment by cartermp** [comment]
This is fixed by #5002
